### PR TITLE
Saml2WebSsoAuthenticationFilter adds authentication details

### PR DESCRIPTION
**Pull request by Kehrlann**
Closes gh-7722.

### CODE REVIEW

**Comment by Kehrlann** [comment]
Hi there!

We've been needing authentication details in our product, specifically `WebAuthenticationDetails`. I saw gh-7722 and decided to submit this.

Notes:
- I added a test in `Saml2WebSsoAuthenticationFilterTests` rather than more integration-style `Saml2LoginConfigurerTests`, but I am unsure whether it is the right place. Maybe they should be in both?
- There is a cast to `AbstractAuthenticationToken`. Technically the authentication converter returns a `Saml2AuthenticationToken` by default. Howver, that is not required. I can't really imagine a setup where a user provides both their own converter AND their own auth provider (they should probably provide their own filter if they do this), but I prefer being extra cautious. Please advise if that is too cautious.

File: saml2/saml2-service-provider/src/main/java/org/springframework/security/saml2/provider/service/servlet/filter/Saml2WebSsoAuthenticationFilter.java [hunk 1]
```diff
@@ -109,6 +110,12 @@ public Authentication attemptAuthentication(HttpServletRequest request, HttpServ
 					"No relying party registration found");
 			throw new Saml2AuthenticationException(saml2Error);
 		}
+
+		if (AbstractAuthenticationToken.class.isAssignableFrom(authentication.getClass())) {
+			Object details = this.authenticationDetailsSource.buildDetails(request);
+			((AbstractAuthenticationToken) authentication).setDetails(details);
+		}
+
```
Comment (marcusdacoregio) [inline saml2/saml2-service-provider/src/main/java/org/springframework/security/saml2/provider/service/servlet/filter/Saml2WebSsoAuthenticationFilter.java:118]: Could you please extract this in a private method called `setDetails` so it becomes more readable?

**Review by marcusdacoregio — CHANGES_REQUESTED** [review]
Hi @Kehrlann, thank you very much for the PR. I've left some feedback inline.

File: saml2/saml2-service-provider/src/test/java/org/springframework/security/saml2/provider/service/servlet/filter/Saml2WebSsoAuthenticationFilterTests.java [hunk 2]
```diff
@@ -119,6 +123,42 @@ public void attemptAuthenticationWhenSavedAuthnRequestThenRemovesAuthnRequest()
 		verify(authenticationRequestRepository).removeAuthenticationRequest(this.request, this.response);
 	}
 
+	@Test
+	public void attemptAuthenticationAddsDetails() {
+		Saml2AuthenticationRequestRepository<AbstractSaml2AuthenticationRequest> authenticationRequestRepository = mock(
```
[the remark this answers is not in the document]
Comment (Kehrlann) [reply saml2/saml2-service-provider/src/test/java/org/springframework/security/saml2/provider/service/servlet/filter/Saml2WebSsoAuthenticationFilterTests.java]: Good catch 👍

File: saml2/saml2-service-provider/src/main/java/org/springframework/security/saml2/provider/service/servlet/filter/Saml2WebSsoAuthenticationFilter.java [hunk 3]
```diff
@@ -109,6 +110,7 @@ public Authentication attemptAuthentication(HttpServletRequest request, HttpServ
 					"No relying party registration found");
 			throw new Saml2AuthenticationException(saml2Error);
 		}
+		this.setDetails(request, authentication);
```
Comment (marcusdacoregio) [inline saml2/saml2-service-provider/src/main/java/org/springframework/security/saml2/provider/service/servlet/filter/Saml2WebSsoAuthenticationFilter.java:113]: In Spring Security's codebase, there are some places that we still call an instance method using `this`, but we are moving away from it. Would you please remove the `this.` and just use `setDetails(request, authentication)`?

File: saml2/saml2-service-provider/src/test/java/org/springframework/security/saml2/provider/service/servlet/filter/Saml2WebSsoAuthenticationFilterTests.java [hunk 4]
```diff
@@ -119,6 +123,39 @@ public void attemptAuthenticationWhenSavedAuthnRequestThenRemovesAuthnRequest()
...
+		this.filter.setAuthenticationDetailsSource(authenticationDetailsSource);
+		this.request.setPathInfo("/some/other/path/idp-registration-id");
+		this.filter.attemptAuthentication(this.request, this.response);
+		Assertions.assertEquals(details, token.getDetails());
+	}
+
+	@Test
+	public void attemptAuthenticationWhenAuthenticationNotAbstractAuthenticationTokenDoesNotAddDetails() {
+		Saml2AuthenticationRequestRepository<AbstractSaml2AuthenticationRequest> authenticationRequestRepository = mock(
+				Saml2AuthenticationRequestRepository.class);
```
Comment (marcusdacoregio) [inline saml2/saml2-service-provider/src/test/java/org/springframework/security/saml2/provider/service/servlet/filter/Saml2WebSsoAuthenticationFilterTests.java:145]: I think the `authenticationRequestRepository` is not needed in this test.

**Review by marcusdacoregio — CHANGES_REQUESTED** [review]
Thank you @Kehrlann.

I've left a couple more suggestions in line. Once you fix them we can merge the PR!

**Review by marcusdacoregio — APPROVED** [review]

**Comment by marcusdacoregio** [comment]
Thank you very much @Kehrlann. This is now merged into `main`.

**Pull request merged**